### PR TITLE
USB Protection Improvement and ExamMode Fix

### PR DESCRIPTION
**Pull request by devdl11**
Some little improvements preparing the release

### CODE REVIEW

File: apps/settings/sub_menu/exam_mode_controller.cpp [hunk 1]
```diff
@@ -112,7 +113,11 @@ GlobalPreferences::ExamMode ExamModeController::examMode() {
   GlobalPreferences::ExamMode mode = GlobalPreferences::sharedGlobalPreferences()->tempExamMode();
   if (GlobalPreferences::sharedGlobalPreferences()->isInExamMode()) {
     // If the exam mode is already on, this re-activate the same exam mode
+    #if ION_SIMULATOR_FILES
+    mode = GlobalPreferences::ExamMode::Off;
+    #else
```
Comment (Yaya-Cout) [inline apps/settings/sub_menu/exam_mode_controller.cpp]: I think these lines should be removed, because this pull request is marked as ready for review

File: apps/apps_container.cpp [hunk 2]
```diff
@@ -411,7 +416,7 @@ void AppsContainer::redrawWindow(bool force) {
 
 void AppsContainer::activateExamMode(GlobalPreferences::ExamMode examMode) {
   assert(examMode != GlobalPreferences::ExamMode::Off && examMode != GlobalPreferences::ExamMode::Unknown);
-  reset();
+  reset(examMode != GlobalPreferences::ExamMode::Standard);
```
Comment (Yaya-Cout) [inline apps/apps_container.cpp:419]: Why did you remove the files if the exam mode isn't standard ?

File: apps/settings/sub_menu/exam_mode_controller.cpp (hunk 1 above)
Comment (devdl11) [reply apps/settings/sub_menu/exam_mode_controller.cpp]: I don't know because it add the possibility to disable the exam mode in the simulator without rebooting

Comment (Yaya-Cout) [reply apps/settings/sub_menu/exam_mode_controller.cpp]: Yes, but it is marked as ready to merge…

File: apps/apps_container.cpp (hunk 2 above)
Comment (devdl11) [reply apps/apps_container.cpp:419]: Because the standard mode is like in France and I don't know what are the modalities for the others countries . Maybe others countries want that the files are deleted in the exam mode ^^

Comment (Yaya-Cout) [reply apps/apps_container.cpp:419]: OK

File: apps/settings/sub_menu/exam_mode_controller.cpp (hunk 1 above)
Comment (devdl11) [reply apps/settings/sub_menu/exam_mode_controller.cpp]: Indeed because I think it can be a good idea to add the possibility to disable the exam mode in simulators ^^ The exam mode in simulator doesn't make any sense in the first place because you can always leave the simulator and open others app so I don't see the problem by adding the option

Comment (Yaya-Cout) [reply apps/settings/sub_menu/exam_mode_controller.cpp]: Yes, but if we look #131, it is built as a simulator, but it runs on the calculator…

Comment (Lauryy06) [reply apps/settings/sub_menu/exam_mode_controller.cpp]: Yes, I agree with Yaya-Cout

File: apps/shared/function.h [hunk 3]
```diff
@@ -78,6 +78,7 @@ class Function : public ExpressionModelHandle {
     }
     bool isActive() const { return m_active; }
     void setActive(bool active) { m_active = active; }
+    void setMyColor(KDColor color) {m_color = color; }
```
Comment (Lauryy06) [inline apps/shared/function.h]:
```suggestion
    void setMyColor(KDColor color) { m_color = color; }
```

Comment (Lauryy06) [reply apps/shared/function.h]: In fact, is it possible to give a better name to the function?

File: ion/src/device/shared/boot/rt0.cpp [hunk 4]
```diff
@@ -245,7 +245,7 @@ void __attribute__((noinline)) start() {
   Ion::Device::Board::initFPU();
 
   /* Call static C++ object constructors
-   * The C++ compiler creates an initialization function for each static object.
+   * The C++ compiler creates an initialization function   for each static object.
```
Comment (Lauryy06) [inline ion/src/device/shared/boot/rt0.cpp]:
```suggestion
   * The C++ compiler creates an initialization function for each static object.
```

File: ion/src/device/shared/usb/dfu_interface.cpp [hunk 5]
```diff
@@ -78,6 +78,7 @@ void DFUInterface::wholeDataSentCallback(SetupPacket *request, uint8_t *transfer
       // Leave DFU routine: Leave DFU, reset device, jump to application code
       leaveDFUAndReset();
     } else if (m_state == State::dfuDNBUSY) {
+      m_state = State::dfuDNBUSY;
```
Comment (Lauryy06) [inline ion/src/device/shared/usb/dfu_interface.cpp]: I don't understand this line...

File: ion/src/device/shared/usb/dfu_interface.cpp [hunk 6]
```diff
@@ -188,7 +189,7 @@ void DFUInterface::changeAddressPointerIfNeeded() {
 
 void DFUInterface::eraseCommand(uint8_t *transferBuffer, uint16_t transferBufferLength) {
   /* We determine whether the commands asks for a mass erase or which sector to
-   * erase. The erase must be done after the next getStatus request. */
+ * erase. The erase must be done after the next getStatus request. */
```
Comment (Lauryy06) [inline ion/src/device/shared/usb/dfu_interface.cpp]: ?

File: ion/src/device/shared/usb/dfu_interface.cpp [hunk 7]
```diff
@@ -152,7 +153,7 @@ bool DFUInterface::processUploadRequest(SetupPacket *request, uint8_t *transferB
   } else {
     /* We decided to never protect Read operation. Else we would have to check
      * here it is not protected before reading. */
-
+
```
Comment (Lauryy06) [inline ion/src/device/shared/usb/dfu_interface.cpp]: ?

File: ion/src/device/shared/usb/dfu_interface.cpp (hunk 6 above)
Comment (Yaya-Cout) [reply ion/src/device/shared/usb/dfu_interface.cpp]: I think it has been squashed when merging, but this fork isn't up-to-date, so it isn't corrected in this repo

File: apps/shared/function.h (hunk 3 above)
Comment (devdl11) [reply apps/shared/function.h]: setGraphColor ?

File: apps/settings/sub_menu/exam_mode_controller.cpp [hunk 8]
```diff
@@ -112,7 +113,11 @@ GlobalPreferences::ExamMode ExamModeController::examMode() {
   GlobalPreferences::ExamMode mode = GlobalPreferences::sharedGlobalPreferences()->tempExamMode();
   if (GlobalPreferences::sharedGlobalPreferences()->isInExamMode()) {
     // If the exam mode is already on, this re-activate the same exam mode
+    // #if ION_SIMULATOR_FILES
+    // mode = GlobalPreferences::ExamMode::Off;
+    // #else
     mode = GlobalPreferences::sharedGlobalPreferences()->examMode();
+    // #endif
```
Comment (Yaya-Cout) [inline apps/settings/sub_menu/exam_mode_controller.cpp:120]: I think these comments can be removed.

File: ion/src/device/shared/usb/dfu_interface.cpp [hunk 9]
```diff
@@ -217,7 +216,7 @@ void DFUInterface::eraseMemoryIfNeeded() {
   if (m_erasePage < 0) {
     return;
   }
-
+
```
Comment (Yaya-Cout) [inline ion/src/device/shared/usb/dfu_interface.cpp:219]: Remove trailing space…
```suggestion

```

**Review by Yaya-Cout — COMMENTED** [review]
Can you fetch upstream, and test with the trash ?

**Comment by Yaya-Cout** [comment]
I think this pull request have to be updated, because #205 remove the USB protection…

**Comment by devdl11** [comment]
I close this pull request because it's too outdated to be merged, a completed rework needs to be made
